### DISCUSSION TITLE
Make it possible to prefill fields in the app

### DISCUSSION THREAD

**Discussion by Attacler** [directus/directus#10232] [Feature Requests]
Basicly the idea of this is to prefill fields when a url has been opened.
So for example:
I have a collection named tasks.
Then i click on a button in a project named Add Task.
Then the add form will open and have the project ID prefilled.

There are more ways this could be used for example:
if you have quotes, you can press a button "Convert to invoice".
then a new page would open with all products/amounts/prices filled into the invoice.

this allows people to make new records more efficient while still allowing them to make changes before saving.

Also another idea if this gets implemented.
Make it possible to add a paramater: closeOnSave.
That way if you saved the record, it will automaticly close the tab (becomes usefull if the link is opened in a new tab, maybe also a reload in the original tab to finish it).

**Comment by benhaynes** [comment]
Love it! I could have sworn that I added a ticket for this a while ago... but I can't find it, so take my 👍 !

This will be huge for Layouts like "Calendar" too, since we can add a "Create Event" button on the day, and it would take you to the Item Form with that date prefilled. My guess is we'd pass this in as a query string, but I defer to @rijkvanzanten.

**Reply by Attacler** [reply]
I can drop in 25€ for someone to make it, i guess trough issue hunter?
Preferably with the closeOnSave option.

**Reply by benhaynes** [reply]
I'll also toss in $30 if someone wants to take this on! Can post on IssueHunt if there's interest.

**Reply by Attacler** [reply]
> I'll also toss in $30 if someone wants to take this on! Can post on IssueHunt if there's interest.

Maybe bonus points if someone makes it selectable for a button?

So then we have 2 parts:
Making the link work
Making a button work that you can select a collection and what fields need to go where.

**Reply by rijkvanzanten** [reply]
If we can figure out how to do it through environment variables (without conflicting with any other potential query params we might want to add to the detail page), we automatically get those 2 parts out of the box I think.

In order to prevent naming conflicts, I think we can rely on something like `?val[title]=Hello%20World`.

The only thing we'll have to be aware of is the limit on the maximums size of a URL, but it sounds like the use cases we design this for are mostly for things like setting a single date field to the selected box on a calendar etc, not full HTML article bodies.

**Reply by benhaynes** [reply]
This has my blessing! Would we ever be able to get around the URL length by having an alternate version that uses headers or something? Either way, I think that query string structure is ideal for what we're trying to achieve. How crazy is the implementation?

**Reply by rijkvanzanten** [reply]
Not from the browser. You can't set headers or really anything from a regular `<a>` tag besides the href with query params. It's a little tricky to implement, as we'd have to hack it in as default values that are send up to the API on save. We don't want to stage them as actual changes, otherwise we'll show the confirm leave dialog without any user changes 🤔

**Comment by danieljwestman** [comment]
Just wanna wake this discussion up again. Any progress or evolving thoughts? 😊

We would also love such functionality.

**Reply by rijkvanzanten** [reply]
No current in-works development, but definitely something we're interested in supporting 🙂

**Reply by av1m** [reply]
I am interested in this feature. I would be at your disposal to develop it.
Would you be interested? If so, can you guide me on your vision?

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by knulpi** [reply]
@rijkvanzanten If you decide on moving forward with this, I would be happy to majorly contribute to the implementation. As this is basically related to some sort of templates for collections, I am highly interested in getting this one in :)

**Reply by rijkvanzanten** [reply]
This feature request is pretty light on details on how this would actually work, so lemme know your thoughts! First step to getting this in would be to discuss the details, and turn this into a proper RFC that we can then implement against 👍🏻 Happy to chat through it here or on Discord (`@rijkvanzanten`) 🚀

**Reply by SeanDylanGoff** [reply]
I implemented this some time ago here:
https://github.com/directus/directus/pull/15678
An additional potentially useful feature is a kind of upsert url that causes an item with a specified item to be created if it doesnt already exist.
We use those two features in combination with a query parameter "autoFocusField" for easy linking to specific items/fields. If it's possible for me to be part of the discussion for this feature i'd appreciate that.

**Comment by pepeday** [comment]
Hey all. Sorry for reviving this but as I realize this never moved forward or was never implemented? Is there anyone who can/is willing to work on this?

**Reply by rijkvanzanten** [reply]
It's an open feature request. We tackle them based on popularity and time availability 🙂

**Comment by AKhateeb** [comment]
It's highly important for me, I'm staying tuned till this will be implemented, thanks community :+1

**Comment by smishnev** [comment]
Hi all, it would be nice to add some kind of hook, and when you open a form, for example when I open a form to create a "article" or something like else, I would like to go to the database, grab some data from there and pre-fill the form.

**Comment by huevoncito** [comment]
Upvote here! This could be as simple allowing fields to be filled with query params. 

`https://directus.domain.com/admin/content/users/+?first_name=Bob&last_name=Barker`

Is there any major technical barrier to this? Happy to help/

**Comment by rafarc21** [comment]
+1 Upvote! This would be very useful for us!

**Comment by av1m** [comment]
I implemented this, here is a pull request #25824

**Reply by Abdallah-Awwad** [reply]
Hi @av1m! Thanks a lot for your PR and for making this feature happen! 😅

Quick question, after clicking “Save and Create New” what's the expected behavior? 🤔
Should the form keep the original prefill parameters for the next item? or it clears it? 
Curious to hear your thoughts!

Personally, I think persisting the params would be really helpful when adding several items of the same type.
